### PR TITLE
Rhull/issue240 coverage feature

### PR DESCRIPTION
**Pull request by RHull-IDM**
This PR adds coverage rate adjustments to tasks in sims. See #240 .

### CODE REVIEW

File: pacehrh/R/pace_coverage_config.R
Comment (MeWu-IDM) [inline pacehrh/R/pace_coverage_config.R:1]: @RHull-IDM the build broke with this file: could you investigate what's wrong? thanks!
```
pace_coverage_config.R
ERROR: unable to collate and parse R files for package 'pacehrh'
```

Comment (RHull-IDM) [reply pacehrh/R/pace_coverage_config.R:1]: Looking into it

**Review by celiot-IDM — COMMENTED** [review]
I've never published a style guide for PACE-HRH code, but I noticed a couple of places where a style guide would help with consistency.

Basically, basic code format (spaces around operators, etc) should agree with whatever RStudio's Reformat Code command does. (It agrees almost to the letter with the tidyverse code format guide).

Thank you for following the camelCase naming convention for variables. (I assume you also noticed that leading dots are used to indicate local function names. R doesn't enforce this kind of lexical scoping, so it's more a hint for the reader. I also assume you picked up on the convention to use PascalCase for the names of public functions. No snake_case.

And while I'm glad for any commenting at all, could you make sure comments begin with correct sentence capitalization?

Final question ... more substantive ... have you written any unit tests?

File: pacehrh/R/pace_task_time.R
Comment (celiot-IDM) [inline pacehrh/R/pace_task_time.R:35]: I have some concerns with this approach.

I totally agree that this is the easiest place in the code to insert this calculation.

However ...

The challenge I'm seeing goes like this. There are two sides to the resource management story: the demand side (how much time it's going to take to supply the various healthcare services), and the supply side (how many resources are available to address the demand). I think it's important to the design of the system that we keep the demand and supply sides separate. Doing the coverage calculation this way conflates the two sides of the equation. It works in an algebra sense, but it's dubious as code architecture.

Let's discuss.

Comment (RHull-IDM) [reply pacehrh/R/pace_task_time.R:35]: This makes sense! As I traced through the code this seemed like the right place for it, but based on your explanation I agree, it's actually kind of a trap. Let me try again and see if it fits better.

**Comment by RHull-IDM** [comment]
@celiot-IDM I will fix the comment cases, and yes, I figured out the meaning of the other syntax and tried to emulate it. Meikang was going to do some of the tests, but now that she's out of town I'll try to cover as much of that as I can. I'm OOO from next Monday through the first part of July so I'm pushing to get this task wrapped up as much as I can before I leave.

File: pacehrh/R/pace_task_time.R
Comment (celiot-IDM) [reply pacehrh/R/pace_task_time.R:35]: image

**Comment by RHull-IDM** [comment]
I think everything is set up now. I haven't had time to add a testthat script for the changes, unfortunately, but Meikang did implement some validation scripts for the inputs. Making the change to the structure of the .xlsx file was a bigger deal than I thought because I had to update all of the test .xlsx files too. Took me a while to figure that out, but I got there eventually.

I'm out from now until July 12th, so feel free to fix anything that isn't working as expected if it needs to be done before I get back, otherwise I will do it when I return.

File: pacehrh/R/pace_coverage_config.R
Comment (MeWu-IDM) [inline pacehrh/R/pace_coverage_config.R:40]: user should get an error instead of continue to silently assuming the coverage here

File: pacehrh/R/pace_coverage_config.R
Comment (MeWu-IDM) [inline pacehrh/R/pace_coverage_config.R:50]: should check the continuity of years here and return an error if some years are missing

File: pacehrh/R/pace_coverage_config.R
Comment (celiot-IDM) [inline pacehrh/R/pace_coverage_config.R:16]: Called this way, loadTable() does *not* do schema validation. I've removed the comment.

File: pacehrh/R/pace_coverage_config.R
Comment (celiot-IDM) [inline pacehrh/R/pace_coverage_config.R:30]: This code generates an ugly error message when the CommonName label is changed to CommonName_xxx:

Loading task coverage rates sheet Coverage
Incorrect coverage rate config header names: (Indicator)Incorrect coverage rate config header names: (CommonName_xxx)Incorrect coverage rate config header names: (Year 0)Incorrect coverage rate config header names: (Year 1)Incorrect coverage rate config header names: (Year 2)Incorrect coverage rate config header names: (Year 3)Incorrect coverage rate config header names: (Year 4)Incorrect coverage rate config header names: (Year 5)Incorrect coverage rate config header names: (Year 6)Incorrect coverage rate config header names: (Year 7)Incorrect coverage rate config header names: (Year 8)Incorrect coverage rate config header names: (Year 9)Incorrect coverage rate config header names: (Year 10)Incorrect coverage rate config header names: (Year 11)Incorrect coverage rate config header names: (Year 12)Incorrect coverage rate config header names: (Year 13)Incorrect coverage rate config header names: (Year 14)Incorrect coverage rate config header names: (Year 15)Incorrect coverage rate config header names: (Year 16)Incorrect coverage rate config header names: (Year 17)Incorrect coverage rate config header names: (Year 18)Incorrect coverage rate config header names: (Year 19)Incorrect coverage rate config header names: (Year 20)

File: pacehrh/R/pace_coverage_config.R
Comment (celiot-IDM) [inline pacehrh/R/pace_coverage_config.R:14]: This line isn't necessary. (Immediately overwritten by loadTable() return value.)

File: pacehrh/R/pace_coverage_config.R
Comment (celiot-IDM) [inline pacehrh/R/pace_coverage_config.R:54]: Is CommonName ever used? Indicator is the unique key for Task records, so there should be no need to carry CommonName around for something that is essentially an internal lookup table.

File: pacehrh/R/pace_cadre_process.R
Comment (celiot-IDM) [inline pacehrh/R/pace_cadre_process.R:263]: I question the decision to replace Service_time with Coverage_service_time in the cadre allocation CSV file instead of  keeping Service_time and adding a Coverage column so users have more flexibility in their later analysis.

(FWIW I see this as part of the broader story of designing for people who aren't us. When we design for a narrow use case - in this case, analysis that takes coverage into account - we're focusing on the supply side of the problem and under-stating the supply side. If we output all three of Service_time, Coverage and Coverage_service_time, users could do analysis from either direction.)

File: pacehrh/R/pace_save_results.R
Comment (celiot-IDM) [inline pacehrh/R/pace_save_results.R:296]: Wow. Well spotted! It's been ages since we've cared about results without seasonality.

**Review by celiot-IDM — APPROVED** [review]
I've completed my review and made a few small changes (some to address Meikang's suggestions).

I have some big-picture concerns, but at this time I think we should be guided by feedback from real use.

Thank you Ryan for your hard work on this!

**Comment by celiot-IDM** [comment]
BTW We've config'ed GitHub to only require *one* approver for a PR merge, but I would like to get positive sign-off from @MeWu-IDM as well as me before we merge.

File: pacehrh/R/pace_cadre_process.R
Comment (MeWu-IDM) [reply pacehrh/R/pace_cadre_process.R:263]: I think we should return both w/wo Coverage as from `pacehrh::SaveExtendedSuiteResults` we can see both columns so it is confusing that it changes after cadre allocation

File: pacehrh/R/pace_coverage_config.R
Comment (MeWu-IDM) [inline pacehrh/R/pace_coverage_config.R:49]: I think it should give user warning about missing years

**Comment by RHull-IDM** [comment]
@celiot-IDM and @MeWu-IDM  Thanks for all the feedback and commits you made to this PR! Sorry I wasn't very responsive last week. My post-vacation brain was unbelieveably fuzzy.

**Comment by celiot-IDM** [comment]
> @celiot-IDM and @MeWu-IDM Thanks for all the feedback and commits you made to this PR! Sorry I wasn't very responsive last week. My post-vacation brain was unbelieveably fuzzy.

You'e very welcome @RHull-IDM! We're a team. I'm glad you got to relax.

File: pacehrh/R/pace_cadre_process.R
Comment (RHull-IDM) [reply pacehrh/R/pace_cadre_process.R:264]: pace_cadre_process.R changed to include both Service_time and Coverage_service_time. Committed and pushed.

**Review by celiot-IDM — APPROVED** [review]
Looks good!

**Review by MeWu-IDM — APPROVED** [review]
The code looks good to me but we should add instruction about coverage Service time in 
pacehrh::ComputeSummaryStats() 
pacehrh::SaveCadreAllocations() 

Also this [PR](https://github.com/InstituteforDiseaseModeling/PACE-HRH/pull/282) has related document update for coverage, @celiot-IDM did you review it? 

We can merge this and bump the version when all documents are in sync with code

**Comment by celiot-IDM** [comment]
> Also this [PR](https://github.com/InstituteforDiseaseModeling/PACE-HRH/pull/282) has related document update for coverage, @celiot-IDM did you review it?

Review done.

**Comment by celiot-IDM** [comment]
> The code looks good to me but we should add instruction about coverage Service time in pacehrh::ComputeSummaryStats() pacehrh::SaveCadreAllocations()

@MeWu-IDM : Do you mean in the main documentation or in the in-line Roxygen comments?

**Comment by MeWu-IDM** [comment]
> > The code looks good to me but we should add instruction about coverage Service time in pacehrh::ComputeSummaryStats() pacehrh::SaveCadreAllocations()
> 
> @MeWu-IDM : Do you mean in the main documentation or in the in-line Roxygen comments?
The current documentation build process is set to only build when version changes (We do this on purpose as it is a good practice to sync code and documentation always)
 I would suggest
1. Merge 282 first 
2. Add Roxygen comments about coverage and update version
Then we would have a release with everything in it.
What do you think?

**Comment by celiot-IDM** [comment]
> > > The code looks good to me but we should add instruction about coverage Service time in pacehrh::ComputeSummaryStats() pacehrh::SaveCadreAllocations()
> > 
> > 
> > @MeWu-IDM : Do you mean in the main documentation or in the in-line Roxygen comments?
> > The current documentation build process is set to only build when version changes (We do this on purpose as it is a good practice to sync code and documentation always)
> > I would suggest
> 
> 1. Merge 282 first
> 2. Add Roxygen comments about coverage and update version
>    Then we would have a release with everything in it.
>    What do you think?

1. There are 3 reviewers attached to #282. I've done my review. I think at least one other reviewer needs to chime in before #282 is merged.

2. AFAIK, there's no documentation of any sort about the file formats created/used by SaveExtendedSuiteResults(), ComputeSummaryStats(), or SaveCadreAllocations(). So it's not just documentation about the Coverage feature that would be needed. 

So an alternative suggestion.

1. Get a second sign-off on #282 and merge it.
2. Bump the version number in #279 (I'll do that now), and merge the PR. This will kick off the doc build to pick up #282.
3. Create a new issue to write documentation about the CSV file formats.

**Pull request merged**
